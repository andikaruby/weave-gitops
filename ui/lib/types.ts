import { Condition, Interval, SourceRefSourceKind } from "./api/core/types.pb";

export enum GrpcErrorCodes {
  Unknown = 2,
  Unauthenticated = 16,
  NotFound = 5,
}

export type RequestError = Error & {
  code?: number;
};

export enum V2Routes {
  Automations = "/automations",
  Sources = "/sources",
  FluxRuntime = "/flux_runtime",
  Kustomization = "/kustomization",
  HelmRelease = "/helm_release",
  HelmRepo = "/helm_repo",
  GitRepo = "/git_repo",

  // Use this to allow for certain components to route to a 404 and still compile.
  // We want to keep certain components around for future use.
  NotImplemented = "/not_implemented",
}

export const WeGONamespace = "flux-system";

export interface Source {
  name?: string;
  namespace?: string;
  type?: SourceRefSourceKind;
  conditions?: Condition[];
  interval?: Interval;
}

export enum AutomationType {
  Kustomization = "Kustomization",
  HelmRelease = "HelmRelease",
}
