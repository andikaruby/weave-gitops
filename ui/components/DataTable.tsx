import {
  FormControl,
  MenuItem,
  Select,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@material-ui/core";
import _ from "lodash";
import * as React from "react";
import styled from "styled-components";
import Button from "./Button";
import Flex from "./Flex";
import Icon, { IconType } from "./Icon";
import Spacer from "./Spacer";
import Text from "./Text";

/** DataTable Properties  */
export interface Props {
  /** CSS MUI Overrides or other styling. */
  className?: string;
  /** A list of objects with two fields: `label`, which is a string representing the column header, and `value`, which can be a string, or a function that extracts the data needed to fill the table cell. */
  fields: {
    label: string;
    displayLabel: string;
    value: string | ((k: any) => string | JSX.Element);
  }[];
  /** A list of data that will be iterated through to create the columns described in `fields`. */
  rows: any[];
  /** A list of strings representing the sortable columns of the table, passed into lodash's `_.sortBy`. */
  sortFields: string[];
  /** an optional list of string widths for each field/column. */
  widths?: string[];
  /** removes bottom pagination bar. */
  disablePagination?: boolean;
  /** array of options for rows per page. Defaults to [25, 50, 75, 100]. */
  paginationOptions?: number[];
}

const EmptyRow = styled(TableRow)<{ colSpan: number }>`
  font-style: italic;
  td {
    text-align: center;
  }
`;

const TableButton = styled(Button)`
  &.MuiButton-root {
    padding: 0;
    margin: 0;
    text-transform: none;
  }
  &.MuiButton-text {
    color: ${(props) => props.theme.colors.neutral30};
    min-width: 0px;
  }
  &.arrow {
    min-width: 0px;
  }
  &.selected {
    color: ${(props) => props.theme.colors.neutral40};
  }
`;

/** Form DataTable */
function UnstyledDataTable({
  className,
  fields,
  rows,
  sortFields,
  widths,
  disablePagination,
  paginationOptions = [25, 50, 75, 100],
}: Props) {
  const [sort, setSort] = React.useState(sortFields[0]);
  const [reverseSort, setReverseSort] = React.useState(false);
  const [pagination, setPagination] = React.useState({
    start: 0,
    length: paginationOptions[0],
  });
  const sorted = _.sortBy(rows, sort);

  if (reverseSort) {
    sorted.reverse();
  }

  type labelProps = { label: string; displayLabel: string };
  function SortableLabel({ label, displayLabel }: labelProps) {
    return (
      <Flex align start>
        <TableButton
          color="inherit"
          variant="text"
          onClick={() => {
            setReverseSort(sort === label ? !reverseSort : false);
            setSort(label);
          }}
        >
          <h2>{displayLabel}</h2>
        </TableButton>
        <Spacer padding="xxs" />
        {sort === label ? (
          <Icon
            type={IconType.ArrowUpwardIcon}
            size="base"
            className={reverseSort ? "upward" : "downward"}
          />
        ) : (
          <div style={{ width: "16px" }} />
        )}
      </Flex>
    );
  }

  const r = [];
  for (
    let i = pagination.start;
    i < pagination.start + pagination.length;
    i++
  ) {
    if (sorted[i]) {
      r.push(
        <TableRow key={i}>
          {_.map(fields, (f, index) => (
            <TableCell style={widths && { width: widths[index] }} key={f.label}>
              <Text>
                {typeof f.value === "function"
                  ? f.value(sorted[i])
                  : sorted[i][f.value]}
              </Text>
            </TableCell>
          ))}
        </TableRow>
      );
    } else {
      break;
    }
  }

  return (
    <div className={className}>
      <TableContainer>
        <Table aria-label="simple table">
          <TableHead>
            <TableRow>
              {_.map(fields, (f, i) => (
                <TableCell style={widths && { width: widths[i] }} key={f.label}>
                  {sortFields.includes(f.label) ? (
                    <SortableLabel
                      label={f.label}
                      displayLabel={f.displayLabel}
                    />
                  ) : (
                    <h2 className="thead">{f.displayLabel}</h2>
                  )}
                </TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {r.length > 0 ? (
              r.map((row) => row)
            ) : (
              <EmptyRow colSpan={fields.length}>
                <TableCell colSpan={fields.length}>
                  <span style={{ fontStyle: "italic" }}>No rows</span>
                </TableCell>
              </EmptyRow>
            )}
          </TableBody>
        </Table>
      </TableContainer>
      {/* pagination row */}
      <Spacer padding="xs" />
      {!disablePagination && (
        <Flex wide align end>
          <FormControl>
            <Flex align>
              <label htmlFor="pagination">Rows Per Page: </label>
              <Spacer padding="xxs" />
              <Select
                id="pagination"
                variant="outlined"
                defaultValue={paginationOptions[0]}
                onChange={(e: React.ChangeEvent<HTMLSelectElement>) => {
                  const newValue = parseInt(e.target.value);
                  setPagination({ start: 0, length: newValue });
                }}
              >
                {paginationOptions.map((option, index) => {
                  return (
                    <MenuItem key={index} value={option}>
                      {option}
                    </MenuItem>
                  );
                })}
              </Select>
            </Flex>
          </FormControl>
          <Spacer padding="base" />
          <Text>
            {pagination.start + 1} - {pagination.start + r.length} out of{" "}
            {rows.length}
          </Text>
          <Spacer padding="base" />
          <Flex>
            <Button
              color="inherit"
              variant="text"
              aria-label="skip to first page"
              disabled={pagination.start === 0}
              onClick={() => setPagination({ ...pagination, start: 0 })}
            >
              <Icon type={IconType.SkipPreviousIcon} size="medium" />
            </Button>
            <Button
              color="inherit"
              variant="text"
              aria-label="back one page"
              disabled={pagination.start === 0}
              onClick={() =>
                setPagination({
                  ...pagination,
                  start: pagination.start - pagination.length,
                })
              }
            >
              <Icon type={IconType.NavigateBeforeIcon} size="medium" />
            </Button>
            <Button
              color="inherit"
              variant="text"
              aria-label="forward one page"
              disabled={pagination.start + pagination.length >= rows.length}
              onClick={() =>
                setPagination({
                  ...pagination,
                  start: pagination.start + pagination.length,
                })
              }
            >
              <Icon type={IconType.NavigateNextIcon} size="medium" />
            </Button>
            <Button
              color="inherit"
              variant="text"
              aria-label="skip to last page"
              disabled={pagination.start + pagination.length >= rows.length}
              onClick={() => {
                let newStart;
                if (rows.length % pagination.length !== 0)
                  newStart = rows.length - (rows.length % pagination.length);
                else newStart = rows.length - pagination.length;
                setPagination({
                  ...pagination,
                  start: newStart,
                });
              }}
            >
              <Icon type={IconType.SkipNextIcon} size="medium" />
            </Button>
          </Flex>
        </Flex>
      )}
    </div>
  );
}

export const DataTable = styled(UnstyledDataTable)`
  h2 {
    margin: 0px;
  }
  .thead {
    color: ${(props) => props.theme.colors.neutral30};
    font-weight: 800;
  }
  .MuiTableRow-root {
    transition: background 0.5s ease-in-out;
  }
  .MuiTableRow-root:not(.MuiTableRow-head):hover {
    background: ${(props) => props.theme.colors.neutral10};
    transition: background 0.5s ease-in-out;
  }
`;

export default DataTable;
