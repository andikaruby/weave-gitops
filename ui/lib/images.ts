/* eslint-disable */
// @ts-ignore
import background from "url:../images/background.svg";
// @ts-ignore
import fancyCardBackgroundBlue from "url:../images/fancy-card-background-blue.svg";
// @ts-ignore
import fancyCardBackgroundOrange from "url:../images/fancy-card-background-orange.svg";
// @ts-ignore
import fancyCardBackground from "url:../images/fancy-card-background.svg";
// @ts-ignore
import SignInBackground from "url:../images/SignInBackground.svg";
// @ts-ignore
import SignInWheel from "url:../images/SignInWheel.svg";
// @ts-ignore
import WeaveLogo from "url:../images/WeaveLogo.svg";

export default {
  SignInWheel,
  SignInBackground,
  WeaveLogo,
  background,
  fancyCardBackground,
  fancyCardBackgroundBlue,
  fancyCardBackgroundOrange,
};
