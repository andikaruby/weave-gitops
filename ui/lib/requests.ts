// TokenRefreshWrapper is a singleton service wrapper calls a user-provided
// token refresh function if a 401 is returned
//
// It only does a single refresh at a time, requests that are made during
// the refresh wait for the refresh to finish first.
export class TokenRefreshWrapper {
  private static refreshPromise: Promise<void> | null = null;
  private static refreshTokenFn: () => Promise<void>;

  static wrap(service: any, refreshTokenFn: () => Promise<void>) {
    this.refreshTokenFn = refreshTokenFn;
    return new Proxy(service, {
      get: (target, propKey) => {
        const origMethod = target[propKey];
        if (typeof origMethod === "function") {
          return (...args: any[]) => {
            return this.makeRequest(origMethod.bind(target, ...args));
          };
        }
        return target[propKey];
      },
    });
  }

  private static getOrInitiateRefresh(): Promise<void> {
    if (!this.refreshPromise) {
      this.refreshPromise = this.refreshTokenFn().finally(() => {
        // Set the promise back to null once the refresh operation is completed
        this.refreshPromise = null;
      });
    }
    return this.refreshPromise;
  }

  private static async makeRequest(fn: any): Promise<any> {
    // Wait for any ongoing refresh, if there is one, to complete
    if (this.refreshPromise) {
      await this.refreshPromise;
    }

    try {
      // Call the function directly
      return await fn();
    } catch (error) {
      // Check for a 401 status code on the HTTPError
      if (error.code === 401) {
        await this.getOrInitiateRefresh();
        // Try the request again
        return this.makeRequest(fn);
      }

      // Rethrow the error if it's not about the token
      throw error;
    }
  }
}
