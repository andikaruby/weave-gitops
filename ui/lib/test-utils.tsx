import { MuiThemeProvider } from "@material-ui/core";
import { createMemoryHistory } from "history";
import _ from "lodash";
import * as React from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import { Router } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import AppContextProvider, { AppProps } from "../contexts/AppContext";
import CoreClientContextProvider from "../contexts/CoreClientContext";
import {
  Core,
  GetChildObjectsRequest,
  GetChildObjectsResponse,
  GetReconciledObjectsRequest,
  GetReconciledObjectsResponse,
  GetVersionRequest,
  GetVersionResponse,
} from "./api/core/core.pb";
import theme, { muiTheme } from "./theme";
import { RequestError } from "./types";

export type CoreOverrides = {
  GetChildObjects?: (req: GetChildObjectsRequest) => GetChildObjectsResponse;
  GetReconciledObjects?: (
    req: GetReconciledObjectsRequest
  ) => GetReconciledObjectsResponse;
  GetVersion?: (req: GetVersionRequest) => GetVersionResponse;
};

export const createCoreMockClient = (
  ovr: CoreOverrides,
  error?: RequestError
): typeof Core => {
  const promisified = _.reduce(
    ovr,
    (result, handlerFn, method) => {
      result[method] = (req) => {
        if (error) {
          return new Promise((_, reject) => reject(error));
        }
        return new Promise((accept) => accept(handlerFn(req) as any));
      };

      return result;
    },
    {}
  );

  return promisified as typeof Core;
};

export function withTheme(element) {
  return (
    <MuiThemeProvider theme={muiTheme}>
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MuiThemeProvider>
  );
}

type TestContextProps = AppProps & { api?: Core };

export function withContext(
  TestComponent,
  url: string,
  { api = {}, ...appProps }: TestContextProps
) {
  const history = createMemoryHistory();
  history.push(url);
  const queryClient = new QueryClient();
  const isElement = React.isValidElement(TestComponent);
  return (
    <Router history={history}>
      <AppContextProvider renderFooter {...appProps}>
        <QueryClientProvider client={queryClient}>
          <CoreClientContextProvider api={api as any}>
            {isElement ? TestComponent : <TestComponent />}
          </CoreClientContextProvider>
        </QueryClientProvider>
      </AppContextProvider>
    </Router>
  );
}
