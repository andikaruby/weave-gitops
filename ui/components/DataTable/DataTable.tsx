import {
  Checkbox,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@material-ui/core";
import _ from "lodash";
import qs from "query-string";
import * as React from "react";
import { useHistory, useLocation } from "react-router-dom";
import styled from "styled-components";
import { ThemeTypes } from "../../contexts/AppContext";
import { SearchedNamespaces } from "../../lib/types";
import { IconButton } from "../Button";
import CheckboxActions from "../CheckboxActions";
import ChipGroup from "../ChipGroup";
import FilterDialog, {
  FilterConfig,
  FilterSelections,
  selectionsToFilters,
} from "../FilterDialog";
import Flex from "../Flex";
import Icon, { IconType } from "../Icon";
import InfoModal from "../InfoModal";
import SearchField from "../SearchField";
import Spacer from "../Spacer";
import Text from "../Text";
import SortableLabel from "./SortableLabel";
import {
  filterRows,
  filterSelectionsToQueryString,
  filterText,
  initialFormState,
  parseFilterStateFromURL,
  sortByField,
  toPairs,
} from "./helpers";
import { Field, FilterState } from "./types";

/** DataTable Properties  */
export interface Props {
  /** The ID of the table. */
  id?: string;
  /** CSS MUI Overrides or other styling. */
  className?: string;
  /** A list of objects with four fields: `label`, which is a string representing the column header, `value`, which can be a string, or a function that extracts the data needed to fill the table cell, and `sortValue`, which customizes your input to the search function */
  fields: Field[];
  /** A list of data that will be iterated through to create the columns described in `fields`. */
  rows?: any[];
  filters?: FilterConfig;
  dialogOpen?: boolean;
  hasCheckboxes?: boolean;
  hideSearchAndFilters?: boolean;
  emptyMessagePlaceholder?: React.ReactNode;
  onColumnHeaderClick?: (field: Field) => void;
  disableSort?: boolean;
  searchedNamespaces?: SearchedNamespaces;
}
//styled components
const EmptyRow = styled(TableRow)<{ colSpan: number }>`
  td {
    text-align: center;
  }
`;

const TopBar = styled(Flex)`
  max-width: 100%;
`;

const IconFlex = styled(Flex)`
  position: relative;
  padding: 0 ${(props) => props.theme.spacing.small};
`;

/** Form DataTable */
function UnstyledDataTable({
  id,
  className,
  fields,
  rows,
  filters,
  hasCheckboxes: checkboxes,
  dialogOpen,
  hideSearchAndFilters,
  emptyMessagePlaceholder,
  onColumnHeaderClick,
  disableSort,
  searchedNamespaces,
}: Props) {
  //URL info
  const history = useHistory();
  const location = useLocation();
  const search = location.search;
  const state = parseFilterStateFromURL(search);

  const [filterDialogOpen, setFilterDialogOpen] = React.useState(dialogOpen);
  const [searchedNamespacesModalOpen, setSearchedNamespacesModalOpen] =
    React.useState(false);
  const [filterState, setFilterState] = React.useState<FilterState>({
    filters: selectionsToFilters(state.initialSelections, filters),
    formState: initialFormState(filters, state.initialSelections),
    textFilters: state.textFilters,
  });

  const handleFilterChange = (sel: FilterSelections) => {
    const filterQuery = filterSelectionsToQueryString(sel);
    history.replace({ ...location, search: filterQuery });
  };

  let filtered = filterRows(rows, filterState.filters);
  filtered = filterText(filtered, fields, filterState.textFilters);
  const chips = toPairs(filterState);

  const doChange = (formState) => {
    if (handleFilterChange) {
      handleFilterChange(formState);
    }
  };

  const handleChipRemove = (chips: string[], filterList) => {
    const next = {
      ...filterState,
    };

    _.each(chips, (chip) => {
      next.formState[chip] = false;
    });

    const filters = selectionsToFilters(next.formState, filterList);

    const textFilters = _.filter(
      next.textFilters,
      (f) => !_.includes(chips, f)
    );

    let query = qs.parse(search);

    if (textFilters.length) query["search"] = textFilters.join("_") + "_";
    else if (query["search"]) query = _.omit(query, "search");
    history.replace({ ...location, search: qs.stringify(query) });

    doChange(next.formState);
    setFilterState({ formState: next.formState, filters, textFilters });
  };

  const handleTextSearchSubmit = (val: string) => {
    if (!val) return;
    const query = qs.parse(search);
    if (!query["search"]) query["search"] = `${val}_`;
    if (!query["search"].includes(val)) query["search"] += `${val}_`;
    history.replace({ ...location, search: qs.stringify(query) });
    setFilterState({
      ...filterState,
      textFilters: _.uniq([...filterState.textFilters, val]),
    });
  };

  const handleClearAll = () => {
    const resetFormState = initialFormState(filters);
    setFilterState({
      filters: {},
      formState: resetFormState,
      textFilters: [],
    });
    const url = qs.parse(location.search);
    //keeps things like clusterName and namespace for details pages
    const cleared = _.omit(url, ["filters", "search"]);
    history.replace({ ...location, search: qs.stringify(cleared) });
  };

  const handleFilterSelect = (filters, formState) => {
    doChange(formState);
    setFilterState({ ...filterState, filters, formState });
  };

  const [sortFieldIndex, setSortFieldIndex] = React.useState(() => {
    let sortFieldIndex = fields.findIndex((f) => f.defaultSort);

    if (sortFieldIndex === -1) {
      sortFieldIndex = 0;
    }

    return sortFieldIndex;
  });

  const secondarySortFieldIndex = fields.findIndex((f) => f.secondarySort);

  const [reverseSort, setReverseSort] = React.useState(false);

  let sortFields = [fields[sortFieldIndex]];

  const useSecondarySort =
    secondarySortFieldIndex > -1 && sortFieldIndex != secondarySortFieldIndex;

  if (useSecondarySort) {
    sortFields = sortFields.concat(fields[secondarySortFieldIndex]);
    sortFields = sortFields.concat(
      fields.filter(
        (_, index) =>
          index != sortFieldIndex && index != secondarySortFieldIndex
      )
    );
  } else {
    sortFields = sortFields.concat(
      fields.filter((_, index) => index != sortFieldIndex)
    );
  }

  const sorted = sortByField(
    filtered,
    reverseSort,
    sortFields,
    useSecondarySort,
    disableSort
  );

  const numFields = fields.length + (checkboxes ? 1 : 0);

  const [checked, setChecked] = React.useState([]);

  const r = _.map(sorted, (r, i) => {
    return (
      <TableRow key={r.uid || i}>
        {checkboxes && (
          <TableCell style={{ padding: "0px" }}>
            <Checkbox
              checked={_.includes(checked, r.uid)}
              onChange={(e) => {
                if (e.target.checked) setChecked([...checked, r.uid]);
                else setChecked(_.without(checked, r.uid));
              }}
              color="primary"
            />
          </TableCell>
        )}
        {_.map(fields, (f) => {
          const style: React.CSSProperties = {
            ...(f.minWidth && { minWidth: f.minWidth }),
            ...(f.maxWidth && { maxWidth: f.maxWidth }),
          };

          return (
            <TableCell
              style={Object.keys(style).length > 0 ? style : undefined}
              key={f.label}
            >
              <Text>
                {(typeof f.value === "function" ? f.value(r) : r[f.value]) ||
                  "-"}
              </Text>
            </TableCell>
          );
        })}
      </TableRow>
    );
  });

  return (
    <Flex wide tall column className={className}>
      <InfoModal
        searchedNamespaces={searchedNamespaces}
        open={searchedNamespacesModalOpen}
        onCloseModal={setSearchedNamespacesModalOpen}
      />
      <TopBar wide align end>
        {checkboxes && <CheckboxActions checked={checked} rows={filtered} />}
        {filters && !hideSearchAndFilters && (
          <>
            <ChipGroup
              chips={chips}
              onChipRemove={(chips) => handleChipRemove(chips, filters)}
              onClearAll={handleClearAll}
            />
            <IconFlex align>
              {searchedNamespaces && (
                <IconButton
                  onClick={() =>
                    setSearchedNamespacesModalOpen(!searchedNamespacesModalOpen)
                  }
                  variant="text"
                >
                  <Icon
                    type={IconType.InfoIcon}
                    size="medium"
                    color="neutral20"
                  />
                </IconButton>
              )}
              <SearchField onSubmit={handleTextSearchSubmit} />
              <IconButton
                onClick={() => setFilterDialogOpen(!filterDialogOpen)}
                variant={filterDialogOpen ? "contained" : "text"}
                color="inherit"
              >
                <Icon
                  type={IconType.FilterIcon}
                  size="medium"
                  color="neutral30"
                />
              </IconButton>
            </IconFlex>
          </>
        )}
      </TopBar>
      <Flex wide tall>
        <TableContainer id={id}>
          <Table aria-label="simple table">
            <TableHead>
              <TableRow>
                {checkboxes && (
                  <TableCell key={"checkboxes"}>
                    <Checkbox
                      checked={filtered?.length === checked.length}
                      onChange={(e) =>
                        e.target.checked
                          ? setChecked(filtered?.map((r) => r.uid))
                          : setChecked([])
                      }
                      color="primary"
                    />
                  </TableCell>
                )}
                {_.map(fields, (f, index) => (
                  <TableCell key={f.label}>
                    {typeof f.labelRenderer === "function" ? (
                      f.labelRenderer(r)
                    ) : (
                      <SortableLabel
                        fields={fields}
                        fieldIndex={index}
                        sortFieldIndex={sortFieldIndex}
                        reverseSort={reverseSort}
                        setSortFieldIndex={(...args) => {
                          if (onColumnHeaderClick) {
                            onColumnHeaderClick(f);
                          }

                          setSortFieldIndex(...args);
                        }}
                        setReverseSort={(isReverse) => {
                          if (onColumnHeaderClick) {
                            onColumnHeaderClick(f);
                          }

                          setReverseSort(isReverse);
                        }}
                      />
                    )}
                  </TableCell>
                ))}
              </TableRow>
            </TableHead>
            <TableBody>
              {r.length > 0 ? (
                r
              ) : (
                <EmptyRow colSpan={numFields}>
                  <TableCell colSpan={numFields}>
                    <Flex center align>
                      <Icon
                        color="neutral20"
                        type={IconType.RemoveCircleIcon}
                        size="base"
                      />
                      <Spacer padding="xxs" />
                      {emptyMessagePlaceholder || (
                        <Text color="neutral30">No data</Text>
                      )}
                    </Flex>
                  </TableCell>
                </EmptyRow>
              )}
            </TableBody>
          </Table>
        </TableContainer>
        {!hideSearchAndFilters && (
          <FilterDialog
            onFilterSelect={handleFilterSelect}
            filterList={filters}
            formState={filterState.formState}
            open={filterDialogOpen}
          />
        )}
      </Flex>
    </Flex>
  );
}
export const DataTable = styled(UnstyledDataTable)`
  width: 100%;
  flex-wrap: nowrap;
  overflow-x: hidden;
  h2 {
    padding: ${(props) => props.theme.spacing.xs};
    font-size: 12px;
    font-weight: 600;
    color: ${(props) => props.theme.colors.neutral30};
    margin: 0px;
    white-space: nowrap;
    text-transform: uppercase;
    letter-spacing: 1px;
  }
  .MuiTableRow-root {
    transition: background 0.5s ease-in-out;
  }
  .MuiTableRow-root:not(.MuiTableRow-head):hover {
    background: ${(props) =>
      props.theme.mode === ThemeTypes.Dark
        ? props.theme.colors.blueWithOpacity
        : props.theme.colors.neutral10};
    transition: background 0.5s ease-in-out;
  }

  .MuiTableCell-root {
    border-color: ${(props) =>
      props.theme.mode === ThemeTypes.Dark
        ? props.theme.colors.primary30
        : props.theme.colors.neutral20};
  }
  table {
    margin-top: ${(props) => props.theme.spacing.small};
  }
  th {
    padding: 0;
    background: ${(props) => props.theme.colors.neutralGray};
    .MuiCheckbox-root {
      padding: 4px 9px;
    }
    :first-child {
      border-top-left-radius: ${(props) => props.theme.spacing.xxs};
    }
    :last-child {
      border-top-right-radius: ${(props) => props.theme.spacing.xxs};
    }
  }
  td {
    //24px matches th + button + h2 padding
    padding-left: ${(props) => props.theme.spacing.base};
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }
  //override so filter dialog button stays highlighted, but color is too bright in dark mode
  .MuiButton-contained {
    ${(props) =>
      props.theme.mode === ThemeTypes.Dark
        ? `background-color: ${props.theme.colors.neutral10};`
        : null}
  }
`;

export default DataTable;
