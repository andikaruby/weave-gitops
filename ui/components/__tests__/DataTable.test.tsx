import { fireEvent, render, screen, within } from "@testing-library/react";
import "jest-styled-components";
import React from "react";
import renderer from "react-test-renderer";
import { withTheme } from "../../lib/test-utils";
import DataTable from "../DataTable";

describe("DataTable", () => {
  const rows = [
    {
      name: "the-cool-app",
      status: "Ready",
      lastUpdate: "2006-01-02T15:04:05-0700",
    },
    {
      name: "podinfo",
      status: "Failed",
      lastUpdate: "2006-01-02T15:04:05-0700",
    },
    {
      name: "nginx",
      status: "Ready",
      lastUpdate: "2006-01-02T15:04:05-0700",
    },
  ];

  const fields = [
    {
      label: "name",
      displayLabel: "Name",
      value: ({ name }) => <a href="/some_url">{name}</a>,
    },
    {
      label: "status",
      displayLabel: "Status",
      value: (v) => v.status,
    },
    {
      label: "lastUpdate",
      displayLabel: "Last Updated",
      value: "lastUpdate",
    },
  ];
  describe("sorting", () => {
    it("initially sorts based on sortFields[0]", () => {
      render(
        withTheme(
          <DataTable
            sortFields={["name", "status"]}
            fields={fields}
            rows={rows}
          />
        )
      );
      const firstRow = screen.getAllByRole("row")[1];
      expect(firstRow.innerHTML).toMatch(/nginx/);
    });
    it("reverses sort on thead click", () => {
      render(
        withTheme(
          <DataTable
            sortFields={["name", "status"]}
            fields={fields}
            rows={rows}
          />
        )
      );

      const nameButton = screen.getByText("Name");
      fireEvent.click(nameButton);
      const firstRow = screen.getAllByRole("row")[1];
      expect(firstRow.innerHTML).toMatch(/the-cool-app/);
    });
    it("resets reverseSort and switches sort column on different thead click", () => {
      render(
        withTheme(
          <DataTable
            sortFields={["name", "status"]}
            fields={fields}
            rows={rows}
          />
        )
      );
      const nameButton = screen.getByText("Name");
      fireEvent.click(nameButton);
      const statusButton = screen.getByText("Status");
      fireEvent.click(statusButton);
      const firstRow = screen.getAllByRole("row")[1];
      expect(firstRow.innerHTML).toMatch(/podinfo/);
    });
  });
  describe("pagination", () => {
    it("displays an initial portion of rows based on paginationOptions[0] prop", () => {
      render(
        withTheme(
          <DataTable
            sortFields={["name", "status"]}
            fields={fields}
            rows={rows}
            paginationOptions={[1, 2]}
          />
        )
      );
      expect(screen.getAllByRole("row").length).toEqual(2);
    });
    it("has functioning navigation buttons and display text", () => {
      render(
        withTheme(
          <DataTable
            sortFields={["name", "status"]}
            fields={fields}
            rows={rows}
            paginationOptions={[1, 2]}
          />
        )
      );
      const back = screen.getByLabelText("back one page");
      const skipBack = screen.getByLabelText("skip to first page");
      const forward = screen.getByLabelText("forward one page");
      const skipForward = screen.getByLabelText("skip to last page");
      const displayText = screen.getByText(/1 - 1 out of 3/);
      fireEvent.click(forward);
      expect(displayText.innerHTML).toContain("2 - 2 out of 3");
      fireEvent.click(back);
      expect(displayText.innerHTML).toContain("1 - 1 out of 3");
      fireEvent.click(skipForward);
      expect(displayText.innerHTML).toContain("3 - 3 out of 3");
      fireEvent.click(skipBack);
      expect(displayText.innerHTML).toContain("1 - 1 out of 3");
    });
    it("disables buttons based on page location", () => {
      render(
        withTheme(
          <DataTable
            sortFields={["name", "status"]}
            fields={fields}
            rows={rows}
            paginationOptions={[1, 2]}
          />
        )
      );
      const back = screen.getByLabelText("back one page");
      const skipBack = screen.getByLabelText("skip to first page");
      const forward = screen.getByLabelText("forward one page");
      const skipForward = screen.getByLabelText("skip to last page");
      expect(back.hasAttribute("disabled")).toBeTruthy();
      expect(skipBack.hasAttribute("disabled")).toBeTruthy();
      fireEvent.click(skipForward);
      expect(forward.hasAttribute("disabled")).toBeTruthy();
      expect(skipForward.hasAttribute("disabled")).toBeTruthy();
    });
    it("resets start index to 0 and changes list length on rows per page select change", () => {
      render(
        withTheme(
          <DataTable
            sortFields={["name", "status"]}
            fields={fields}
            rows={rows}
            paginationOptions={[1, 2]}
          />
        )
      );
      const nextPage = screen.getByLabelText("forward one page");
      fireEvent.click(nextPage);
      expect(screen.getAllByRole("row")[1].innerHTML).toMatch(/podinfo/);
      let select;
      screen.getAllByRole("button").forEach((button) => {
        if (button.getAttribute("aria-haspopup")) select = button;
      });
      fireEvent.mouseDown(select);
      const listbox = within(screen.getByRole("listbox"));
      fireEvent.click(listbox.getByText("2"));
      expect(screen.getAllByRole("row")[2].innerHTML).toMatch(/podinfo/);
    });
  });
  describe("snapshots", () => {
    it("renders", () => {
      const tree = renderer
        .create(
          withTheme(
            <DataTable sortFields={["name"]} fields={fields} rows={rows} />
          )
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
