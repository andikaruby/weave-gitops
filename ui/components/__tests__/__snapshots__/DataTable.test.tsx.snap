// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable snapshots renders 1`] = `
<<<<<<< HEAD
=======
.c1 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 16px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
}

.c0 .MuiTableCell-head {
  font-weight: 800;
}

>>>>>>> main
<div
  className=""
>
  <div
    className="MuiTableContainer-root"
  >
    <table
      aria-label="simple table"
      className="MuiTable-root"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            Name
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            Status
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            Last Updated
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            <a
              href="/some_url"
            >
              my-cool-app
            </a>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            Ready
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            2006-01-02T15:04:05-0700
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            <a
              href="/some_url"
            >
              nginx
            </a>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            Ready
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            2006-01-02T15:04:05-0700
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            <a
              href="/some_url"
            >
              podinfo
            </a>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            Ready
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            2006-01-02T15:04:05-0700
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
